File: python/src/swap.py
from enum import Enum
from src.utils import (
    find_case_insensitive_index_in_list,
    _to_scaled_18_apply_rate_round_down,
    _to_scaled_18_apply_rate_round_up,
    _to_raw_undo_rate_round_down,
    _to_raw_undo_rate_round_up,
    _compute_and_charge_aggregate_swap_fees,
)
from src.maths import mul_up_fixed


class SwapKind(Enum):
    GIVENIN = 0
    GIVENOUT = 1


def swap(swap_input, pool_state, pool_class, hook_class, hook_state):

    input_index = find_case_insensitive_index_in_list(
        pool_state["tokens"], swap_input["token_in"]
    )
    if input_index == -1:
        raise SystemError("Input token not found on pool")

    output_index = find_case_insensitive_index_in_list(
        pool_state["tokens"], swap_input["token_out"]
    )
    if input_index == -1:
        raise SystemError("Output token not found on pool")

    amount_given_scaled18 = _update_amount_given_in_vars(
        swap_input["amount_raw"],
        swap_input["swap_kind"],
        input_index,
        output_index,
        pool_state["scalingFactors"],
        pool_state["tokenRates"],
    )

    updated_balances_live_scaled18 = pool_state["balancesLiveScaled18"][:]
    if hook_class.shouldCallBeforeSwap:
        # Note - in SC balances and amounts are updated to reflect any rate change.
        # Daniel said we should not worry about this as any large rate changes
        # will mean something has gone wrong.
        # We do take into account and balance changes due
        # to hook using hookAdjustedBalancesScaled18.
        hook_return = hook_class.onBeforeSwap({**swap_input, "hook_state": hook_state})
        if hook_return["success"] is False:
            raise SystemError("BeforeSwapHookFailed")
        for i, a in enumerate(hook_return["hookAdjustedBalancesScaled18"]):
            updated_balances_live_scaled18[i] = a

    swap_fee = pool_state["swapFee"]
    if hook_class.shouldCallComputeDynamicSwapFee:
        hook_return = hook_class.onComputeDynamicSwapFee(
            swap_input,
            pool_state["swapFee"],
            hook_state,
        )
        if hook_return["success"] is True:
            swap_fee = hook_return["dynamicSwapFee"]

    # _swap()
    swap_params = {
        "swap_kind": swap_input["swap_kind"],
        "amount_given_scaled18": amount_given_scaled18,
        "balances_live_scaled18": updated_balances_live_scaled18,
        "index_in": input_index,
        "index_out": output_index,
    }

    amount_calculated_scaled18 = pool_class.on_swap(swap_params)

    # Set swap_fee_amount_scaled18 based on the amountCalculated.
    swap_fee_amount_scaled18 = 0
    if swap_fee > 0:
        # Swap fee is always a percentage of the amountCalculated.
        # On ExactIn, subtract it from the calculated
        # amountOut. On ExactOut, add it to the calculated amountIn.
        # Round up to avoid losses during precision loss.
        swap_fee_amount_scaled18 = mul_up_fixed(amount_calculated_scaled18, swap_fee)

    amount_calculated_raw = 0
    if swap_input["swap_kind"] == SwapKind.GIVENIN.value:
        amount_calculated_scaled18 -= swap_fee_amount_scaled18
        # For `ExactIn` the amount calculated is leaving the Vault, so we round down.
        amount_calculated_raw = _to_raw_undo_rate_round_down(
            amount_calculated_scaled18,
            pool_state["scalingFactors"][output_index],
            pool_state["tokenRates"][output_index],
        )
    else:
        amount_calculated_scaled18 += swap_fee_amount_scaled18
        # For `ExactOut` the amount calculated is entering the Vault, so we round up.
        amount_calculated_raw = _to_raw_undo_rate_round_up(
            amount_calculated_scaled18,
            pool_state["scalingFactors"][input_index],
            pool_state["tokenRates"][input_index],
        )

    aggregate_swap_fee_amount_scaled18 = _compute_and_charge_aggregate_swap_fees(
        swap_fee_amount_scaled18,
        pool_state["aggregateSwapFee"],
    )

    # For ExactIn, we increase the tokenIn balance by `amountIn`,
    # and decrease the tokenOut balance by the
    # (`amountOut` + fees).
    # For ExactOut, we increase the tokenInBalance by (`amountIn` - fees),
    # and decrease the tokenOut balance by
    # `amountOut`.
    balance_in_increment, balance_out_decrement = (
        (
            amount_given_scaled18,
            amount_calculated_scaled18 + aggregate_swap_fee_amount_scaled18,
        )
        if swap_input["swap_kind"] == SwapKind.GIVENIN
        else (
            amount_calculated_scaled18 - aggregate_swap_fee_amount_scaled18,
            amount_given_scaled18,
        )
    )

    updated_balances_live_scaled18[input_index] += balance_in_increment
    updated_balances_live_scaled18[output_index] -= balance_out_decrement

    if hook_class.shouldCallAfterSwap:
        hook_return = hook_class.onAfterSwap(
            {
                "kind": swap_input["swap_kind"],
                "token_in": swap_input["token_in"],
                "token_out": swap_input["token_out"],
                "amount_in_scaled18": (
                    amount_given_scaled18
                    if swap_input["swap_kind"] == SwapKind.GIVENIN
                    else amount_calculated_scaled18
                ),
                "amount_out_scaled18": (
                    amount_calculated_scaled18
                    if swap_input["swap_kind"] == SwapKind.GIVENIN
                    else amount_given_scaled18
                ),
                "token_in_balance_scaled18": updated_balances_live_scaled18[
                    input_index
                ],
                "token_out_balance_scaled18": updated_balances_live_scaled18[
                    output_index
                ],
                "amount_calculated_scaled18": amount_calculated_scaled18,
                "amount_calculated_raw": amount_calculated_raw,
                "hook_state": hook_state,
            }
        )
        if hook_return["success"] is False:
            raise SystemError(
                "AfterAddSwapHookFailed", pool_state["poolType"], pool_state["hookType"]
            )
        # If hook adjusted amounts is not enabled, ignore amount returned by the hook
        if hook_class.enable_hook_adjusted_amounts:
            amount_calculated_raw = hook_return["hook_adjusted_amount_calculated_raw"]

    return amount_calculated_raw


def _update_amount_given_in_vars(
    amount_given_raw: int,
    swap_kind: SwapKind,
    index_in: int,
    index_out: int,
    scaling_factors: list[int],
    token_rates: list[int],
) -> int:
    # If the amountGiven is entering the pool math (ExactIn), round down
    # since a lower apparent amountIn leads
    # to a lower calculated amountOut, favoring the pool.
    if swap_kind == SwapKind.GIVENIN.value:
        amount_given_scaled_18 = _to_scaled_18_apply_rate_round_down(
            amount_given_raw,
            scaling_factors[index_in],
            token_rates[index_in],
        )
    else:
        amount_given_scaled_18 = _to_scaled_18_apply_rate_round_up(
            amount_given_raw,
            scaling_factors[index_out],
            token_rates[index_out],
        )

    return amount_given_scaled_18

File: rust/src/vault/swap.rs
//! Vault swap operations

use crate::common::constants::WAD;
use crate::common::errors::PoolError;
use crate::common::maths::{complement_fixed, mul_div_up_fixed, mul_up_fixed};
use crate::common::pool_base::PoolBase;
use crate::common::types::*;
use crate::common::utils::{
    compute_and_charge_aggregate_swap_fees, find_case_insensitive_index_in_list, to_raw_undo_rate_round_down, to_raw_undo_rate_round_up, to_scaled_18_apply_rate_round_down, to_scaled_18_apply_rate_round_up,
};
use crate::hooks::types::{AfterSwapParams, HookState};
use crate::hooks::HookBase;
use lazy_static::lazy_static;
use num_bigint::BigInt;
use num_traits::Zero;

lazy_static! {
    /// Minimum trade amount (scaled 18)
    static ref MINIMUM_TRADE_AMOUNT: BigInt = BigInt::from(1000000i64); // 1e6
}

/// Perform a swap operation
pub fn swap(
    swap_input: &SwapInput,
    pool_state: &PoolState,
    pool_class: &dyn PoolBase,
    hook_class: &dyn HookBase,
    hook_state: Option<&HookState>,
) -> Result<BigInt, PoolError> {
    if swap_input.amount_raw.is_zero() {
        return Ok(BigInt::zero());
    }

    let base_state = pool_state.base();

    // Find token indices (case insensitive)
    let input_index = find_case_insensitive_index_in_list(&base_state.tokens, &swap_input.token_in)
        .ok_or(PoolError::InputTokenNotFound)?;

    let output_index = find_case_insensitive_index_in_list(&base_state.tokens, &swap_input.token_out)
        .ok_or(PoolError::OutputTokenNotFound)?;

    // Compute amount given scaled to 18 decimals
    let amount_given_scaled_18 = compute_amount_given_scaled_18(
        &swap_input.amount_raw,
        swap_input.swap_kind.clone(),
        input_index,
        output_index,
        &base_state.scaling_factors,
        &base_state.token_rates,
    )?;

    // Create updated balances
    let mut updated_balances = base_state.balances_live_scaled_18.clone();

    // Create swap parameters
    let mut swap_params = SwapParams {
        swap_kind: swap_input.swap_kind.clone(),
        token_in_index: input_index,
        token_out_index: output_index,
        amount_scaled_18: amount_given_scaled_18.clone(),
        balances_live_scaled_18: updated_balances.clone(),
    };

    // Call before swap hook if needed
    if hook_class.config().should_call_before_swap {
        let result = hook_class.on_before_swap(&swap_params, hook_state.unwrap());
        if !result.success {
            return Err(PoolError::BeforeSwapHookFailed);
        }
        // Update balances with hook-adjusted balances
        for (i, adjusted_balance) in result.hook_adjusted_balances_scaled_18.iter().enumerate() {
            updated_balances[i] = adjusted_balance.clone();
        }
        swap_params.balances_live_scaled_18 = updated_balances.clone();
    }

    // Apply swap fees
    let mut swap_fee = base_state.swap_fee.clone();
    if hook_class.config().should_call_compute_dynamic_swap_fee {
        let result =
            hook_class.on_compute_dynamic_swap_fee(&swap_params, &swap_fee, hook_state.unwrap());
        if result.success {
            swap_fee = result.dynamic_swap_fee;
        }
    }

    let mut total_swap_fee_amount_scaled_18 = BigInt::zero();
    if swap_params.swap_kind == SwapKind::GivenIn {
        // Round up to avoid losses during precision loss
        total_swap_fee_amount_scaled_18 = mul_up_fixed(&swap_params.amount_scaled_18, &swap_fee)?;
        swap_params.amount_scaled_18 =
            &swap_params.amount_scaled_18 - &total_swap_fee_amount_scaled_18;
    }

    ensure_valid_swap_amount(&swap_params.amount_scaled_18)?;

    // Perform the swap
    let amount_calculated_scaled_18 = pool_class.on_swap(&swap_params)?;

    ensure_valid_swap_amount(&amount_calculated_scaled_18)?;

    // Convert result back to raw amount
    let amount_calculated_raw = match swap_input.swap_kind {
        SwapKind::GivenIn => {
            // For ExactIn the amount calculated is leaving the Vault, so we round down
            let rate_rounded_up = compute_rate_round_up(&base_state.token_rates[output_index]);
            to_raw_undo_rate_round_down(
                &amount_calculated_scaled_18,
                &base_state.scaling_factors[output_index],
                &rate_rounded_up,
            )?
        }
        SwapKind::GivenOut => {
            // For ExactOut, add swap fee and round up
            total_swap_fee_amount_scaled_18 = mul_div_up_fixed(
                &amount_calculated_scaled_18,
                &swap_fee,
                &complement_fixed(&swap_fee)?,
            )?;
            let amount_with_fee = &amount_calculated_scaled_18 + &total_swap_fee_amount_scaled_18;

            // For ExactOut the amount calculated is entering the Vault, so we round up
            to_raw_undo_rate_round_up(
                &amount_with_fee,
                &base_state.scaling_factors[input_index],
                &base_state.token_rates[input_index],
            )?
        }
    };

    // Compute and charge aggregate swap fees
    let aggregate_swap_fee_amount_scaled_18 = compute_and_charge_aggregate_swap_fees(
        &total_swap_fee_amount_scaled_18,
        &base_state.aggregate_swap_fee,
        &base_state.scaling_factors,
        &base_state.token_rates,
        input_index,
    )?;

    // Update balances
    let (balance_in_increment, balance_out_decrement) = match swap_input.swap_kind {
        SwapKind::GivenIn => (
            &amount_given_scaled_18 - &aggregate_swap_fee_amount_scaled_18,
            amount_calculated_scaled_18.clone(),
        ),
        SwapKind::GivenOut => (
            &amount_calculated_scaled_18 - &aggregate_swap_fee_amount_scaled_18,
            amount_given_scaled_18.clone(),
        ),
    };

    updated_balances[input_index] = &updated_balances[input_index] + &balance_in_increment;
    updated_balances[output_index] = &updated_balances[output_index] - &balance_out_decrement;

    // Call after swap hook if needed
    let mut final_amount_calculated_raw = amount_calculated_raw.clone();
    if hook_class.config().should_call_after_swap {
        let after_swap_params = AfterSwapParams {
            kind: swap_input.swap_kind.clone(),
            token_in: swap_input.token_in.clone(),
            token_out: swap_input.token_out.clone(),
            amount_in_scaled_18: amount_given_scaled_18.clone(),
            amount_out_scaled_18: amount_calculated_scaled_18.clone(),
            token_in_balance_scaled_18: updated_balances[input_index].clone(),
            token_out_balance_scaled_18: updated_balances[output_index].clone(),
            amount_calculated_scaled_18: amount_calculated_scaled_18.clone(),
            amount_calculated_raw: amount_calculated_raw.clone(),
        };

        let result = hook_class.on_after_swap(&after_swap_params, hook_state.unwrap());
        if !result.success {
            return Err(PoolError::AfterSwapHookFailed);
        }

        // If hook adjusted amounts is enabled, use the hook-adjusted amount
        if hook_class.config().enable_hook_adjusted_amounts {
            final_amount_calculated_raw = result.hook_adjusted_amount_calculated_raw;
        }
    }

    Ok(final_amount_calculated_raw)
}

/// Compute amount given scaled to 18 decimals
pub fn compute_amount_given_scaled_18(
    amount_given_raw: &BigInt,
    swap_kind: SwapKind,
    index_in: usize,
    index_out: usize,
    scaling_factors: &[BigInt],
    token_rates: &[BigInt],
) -> Result<BigInt, PoolError> {
    match swap_kind {
        SwapKind::GivenIn => {
            // If the amountGiven is entering the pool math (ExactIn), round down
            // since a lower apparent amountIn leads to a lower calculated amountOut, favoring the pool.
            Ok(to_scaled_18_apply_rate_round_down(
                amount_given_raw,
                &scaling_factors[index_in],
                &token_rates[index_in],
            )?)
        }
        SwapKind::GivenOut => {
            // For ExactOut, round up to favor the pool
            Ok(to_scaled_18_apply_rate_round_up(
                amount_given_raw,
                &scaling_factors[index_out],
                &token_rates[index_out],
            )?)
        }
    }
}

/// Compute rate rounded up
pub fn compute_rate_round_up(rate: &BigInt) -> BigInt {
    let rounded_rate = (rate / &*WAD) * &*WAD;
    if &rounded_rate == rate {
        rate.clone()
    } else {
        rate + BigInt::from(1)
    }
}

/// Ensure valid swap amount
pub fn ensure_valid_swap_amount(trade_amount: &BigInt) -> Result<(), PoolError> {
    if trade_amount < &*MINIMUM_TRADE_AMOUNT {
        return Err(PoolError::TradeAmountTooSmall);
    }
    Ok(())
}
